README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

powerlmm
========

[![Travis-CI Build
Status](https://travis-ci.org/rpsychologist/powerlmm.svg?branch=master)](https://travis-ci.org/rpsychologist/powerlmm)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/powerlmm)](https://cran.r-project.org/package=powerlmm)
[![](https://cranlogs.r-pkg.org/badges/grand-total/powerlmm)](https://cran.r-project.org/package=powerlmm)

Power Analysis for Longitudinal Multilevel/Linear Mixed-Effects Models.

Overview
--------

The purpose of `powerlmm` is to help design longitudinal treatment
studies (parallel groups), with or without higher-level clustering
(e.g. longitudinally clustered by therapists, groups, or physician), and
missing data. The main features of the package are:

-   Longitudinal two- and three-level (nested) linear mixed-effects
    models, and partially nested designs.
-   Random slopes at the subject- and cluster-level.
-   Missing data.
-   Unbalanced designs (both unequal cluster sizes, and treatment
    groups).
-   Design effect, and estimated type I error when the third-level is
    ignored.
-   Fast analytical power calculations for all designs.
-   Power for small samples sizes using Satterthwaite’s degrees of
    freedom approximation.
-   Explore bias, Type I errors, model misspecification, and LRT model
    selection using convenient simulation methods.

Installation
------------

`powerlmm` can be installed from <sub>CRAN</sub> and GitHub.

``` r
# CRAN, version 0.4.0
# install.packages("powerlmm")

# GitHub, dev version
devtools::install_github("rpsychologist/powerlmm")
```

Example usage
-------------

This is an example of setting up a three-level longitudinal model with
random slopes at both the subject- and cluster-level, with different
missing data patterns per treatment arm. Relative standardized inputs
are used, but unstandardized raw parameters values can also be used.

``` r
library(powerlmm)
d <- per_treatment(
  control = dropout_weibull(0.3, 2),
  treatment = dropout_weibull(0.2, 2)
)
p <- study_parameters(
  n1 = 11,
  n2 = 10,
  n3 = 5,
  T_end = 10,
  icc_pre_subject = 0.5,
  icc_pre_cluster = 0,
  icc_slope = 0.05,
  var_ratio = 0.02,
  dropout = d,
  effect_size = cohend(
    ES = -0.8,
    standardizer = "pretest_SD"
  )
)
p
#> 
#>      Study setup (three-level, nested) 
#> 
#>               n1 = 11
#>               n2 = 10 x 5 (treatment)
#>                    10 x 5 (control)
#>               n3 = 5      (treatment)
#>                    5      (control)
#>                    10     (total)
#>          total_n = 50     (treatment)
#>                    50     (control)
#>                    100    (total)
#>          dropout =  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10 (time)
#>                     0,  0,  1,  3,  6,  9, 12, 16, 20, 25, 30 (%, control)
#>                     0,  0,  1,  2,  4,  5,  8, 10, 13, 17, 20 (%, treatment)
#> icc_pre_subjects = 0.5
#> icc_pre_clusters = 0
#>        icc_slope = 0.05
#>        var_ratio = 0.02
#>      effect_size = -0.8 (Cohen's d [SD: pretest_SD])
```

``` r
plot(p)
```

<img src="man/figures/README-trend-1.png" style="display: block; margin: auto;" />

``` r
get_power(p, df = "satterthwaite")
#> NULL
#> 
#>      Power Analysis for Longitudinal Linear Mixed-Effects Models (3-level,)
#>                   with missing data and unbalanced designs 
#> 
#>               n1 = 11
#>               n2 = 10 x 5 (treatment)
#>                    10 x 5 (control)
#>               n3 = 5      (treatment)
#>                    5      (control)
#>                    10     (total)
#>          total_n = 50     (control)
#>                    50     (treatment)
#>                    100    (total)
#>          dropout =  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10 (time)
#>                     0,  0,  1,  3,  6,  9, 12, 16, 20, 25, 30 (%, control)
#>                     0,  0,  1,  2,  4,  5,  8, 10, 13, 17, 20 (%, treatment)
#> icc_pre_subjects = 0.5
#> icc_pre_clusters = 0
#>        icc_slope = 0.05
#>        var_ratio = 0.02
#>      effect_size = -0.8 (Cohen's d [SD: pretest_SD])
#>               df = 7.957674
#>            alpha = 0.05
#>            power = 68%
```

More vizualisations
-------------------

``` r
plot(
  p,
  RE_level = c(1, 2),
  type = "trend"
)
```

<img src="man/figures/README-trends-varying-1.png" style="display: block; margin: auto;" />

``` r
plot(
  update(p, n1 = 3),
  RE_level = c(1, 2),
  type = "trend_ridges"
)
#> Warning in FUN(X[[i]], ...): no non-missing arguments to max; returning -Inf

#> Warning in FUN(X[[i]], ...): no non-missing arguments to max; returning -Inf
```

<img src="man/figures/README-trend-ridges-1.png" style="display: block; margin: auto;" />

### Unequal cluster sizes

Unequal cluster sizes is also supported, the cluster sizes can either be
random (sampled), or the marginal distribution can be specified.

``` r
p <- study_parameters(
  n1 = 11,
  n2 = unequal_clusters(2, 3, 5, 20),
  icc_pre_subject = 0.5,
  icc_pre_cluster = 0,
  icc_slope = 0.05,
  var_ratio = 0.02,
  effect_size = cohend(-0.8,
    standardizer = "pretest_SD")
)
get_power(p)
#> NULL
#> 
#>      Power Analysis for Longitudinal Linear Mixed-Effects Models (3-level,)
#>                   with missing data and unbalanced designs 
#> 
#>               n1 = 11
#>               n2 = 2, 3, 5, 20 (treatment)
#>                    2, 3, 5, 20 (control)
#>               n3 = 4           (treatment)
#>                    4           (control)
#>                    8           (total)
#>          total_n = 30          (control)
#>                    30          (treatment)
#>                    60          (total)
#>          dropout = No missing data
#> icc_pre_subjects = 0.5
#> icc_pre_clusters = 0
#>        icc_slope = 0.05
#>        var_ratio = 0.02
#>      effect_size = -0.8 (Cohen's d [SD: pretest_SD])
#>               df = 6
#>            alpha = 0.05
#>            power = 44%
```

Cluster sizes follow a Poisson distribution

``` r
p <- study_parameters(
  n1 = 11,
  n2 = unequal_clusters(func = rpois(5, 5)), # sample from Poisson
  icc_pre_subject = 0.5,
  icc_pre_cluster = 0,
  icc_slope = 0.05,
  var_ratio = 0.02,
  effect_size = cohend(
    ES = -0.8,
    standardizer = "pretest_SD"
  )
)
# expected power by averaging over R realizations
get_power(p, R = 100, progress = FALSE)
#> NULL
#> 
#>      Power Analysis for Longitudinal Linear Mixed-Effects Models (3-level,)
#>                   with missing data and unbalanced designs 
#> 
#>               n1 = 11
#>               n2 = rpois(5, 5) (treatment)
#>                    -           (control)
#>               n3 = 5           (treatment)
#>                    5           (control)
#>                    10          (total)
#>          total_n = 24.67       (control)
#>                    24.67       (treatment)
#>                    49.34       (total)
#>          dropout = No missing data
#> icc_pre_subjects = 0.5
#> icc_pre_clusters = 0
#>        icc_slope = 0.05
#>        var_ratio = 0.02
#>      effect_size = -0.8 (Cohen's d [SD: pretest_SD])
#>               df = 8
#>            alpha = 0.05
#>            power = 47% (MCSE: 1%)
#> 
#> NOTE: n2 is randomly sampled. Values are the mean from R = 100 realizations.
```

### Convenience functions

Several convenience functions are also included, e.g. for creating power
curves.

``` r
x <- get_power_table(
  p,
  n2 = 5:10,
  n3 = c(4, 8, 12),
  effect_size = cohend(
    ES = c(0.5, 0.8),
    standardizer = "pretest_SD"
  ),
  cores = CORES
)
```

``` r
plot(x)
```

<img src="man/figures/README-power-curve-1.png" style="display: block; margin: auto;" />

Simulation
----------

The package includes a flexible simulation method that makes it easy to
investigate the performance of different models. As an example, let’s
compare the power difference between the 2-level LMM with 11 repeated
measures, to doing an ANCOVA at posttest. Using `sim_formula` different
models can be fit to the same data set during the simulation.

``` r
p <- study_parameters(
  n1 = 11,
  n2 = 40,
  icc_pre_subject = 0.5,
  cor_subject = -0.4,
  var_ratio = 0.02,
  effect_size = cohend(
    ES = -0.8,
    standardizer = "pretest_SD"
  )
)
# 2-lvl LMM
f0 <- sim_formula("y ~ time + time:treatment + (1 + time | subject)")
# ANCOVA, formulas with no random effects are with using lm()
f1 <- sim_formula("y ~ treatment + pretest",
  data_transform = transform_to_posttest,
  test = "treatment")
f <- sim_formula_compare("LMM" = f0,
  "ANCOVA" = f1)
res <- simulate(
  p,
  nsim = 2000,
  formula = f,
  cores = CORES
)
```

We then summarize the results using `summary`. Let’s look specifically
at the treatment effects.

``` r
summary(
  res,
  para = list(
    "LMM" = "time:treatment",
    "ANCOVA" = "treatment"
  )
)
#> Model:  summary 
#> 
#> Fixed effects: 'time:treatment', 'treatment'
#> 
#>   model M_est theta M_se SD_est Power Power_bw Power_satt
#>     LMM  -1.1  -1.1 0.32   0.33  0.93     0.93        NaN
#>  ANCOVA -11.0   0.0 3.70   3.80  0.83     0.82       0.82
#> ---
#> Number of simulations: 2000  | alpha:  0.05
#> Time points (n1):  11
#> Subjects per cluster (n2 x n3):  40 (treatment)
#>                                  40 (control)
#> Total number of subjects:  80 
#> ---
#> At least one of the models applied a data transformation during simulation,
#> summaries that depend on the true parameter values will no longer be correct,
#> see 'help(summary.plcp_sim)'
```

We can also look at a specific model, here’s the results for the 2-lvl
LMM.

``` r
summary(res, model = "LMM")
#> Model:  LMM 
#> 
#> Random effects 
#> 
#>          parameter  M_est theta est_rel_bias prop_zero is_NA
#>  subject_intercept 100.00 100.0         0.01         0     0
#>      subject_slope   2.00   2.0        -0.01         0     0
#>              error 100.00 100.0         0.00         0     0
#>        cor_subject  -0.39  -0.4        -0.03         0     0
#> 
#> Fixed effects 
#> 
#>       parameter M_est theta M_se SD_est Power Power_bw Power_satt
#>     (Intercept) -0.10   0.0 1.30   1.20  0.03        .        NaN
#>            time -0.02   0.0 0.25   0.25  0.05        .        NaN
#>  time:treatment -1.10  -1.1 0.32   0.33  0.93     0.93        NaN
#> ---
#> Number of simulations: 2000  | alpha:  0.05
#> Time points (n1):  11
#> Subjects per cluster (n2 x n3):  40 (treatment)
#>                                  40 (control)
#> Total number of subjects:  80
#> [Model: LMM] 15.1% of the models threw convergence warnings
#> ---
#> At least one of the models applied a data transformation during simulation,
#> summaries that depend on the true parameter values will no longer be correct,
#> see 'help(summary.plcp_sim)'
```

Launch interactive web application
----------------------------------

The package’s basic functionality is also implemented in a Shiny web
application, aimed at users that are less familiar with R. Launch the
application by typing

``` r
library(powerlmm)
shiny_powerlmm()
```

![](man/figures/README-shiny-screenshot1.png)

![](man/figures/README-shiny-screenshot1.png)


NEWS.md


# Changes in version 0.5.0

## Minor changes
* `simulate(..., seed = 1234)` no longer ignores the seed argument. If cores > 1 the "L'Ecuyer-CMRG" RNG is used via `parallel::clusterSetRNGStream`.

## Bug fixes
* `Summary(model = ...)` now correctly handles a vector of model names, and will give more informative error messages.

# Changes in version 0.4.0

This version substantially improves the `simulate` method.

## New features
* The simulate method is now much more flexible. New features include:
   - Compare more than 2 model formulas (#2).
   - Apply a data transformation during simulation; `sim_formula(..., data_transform = func)`.
     As an example `transform_to_posttest` is included.
   - Choose which parameters to test; `sim_formula(..., test = "treatment")`
   - Fit OLS models. If a model formula is supplied that contain no random effects the
   model is fit using OLS `lm()`. If this is combined with the `transform_to_posttest` the
   longitudinal model can be compared to a cross-sectional model, e.g. ANCOVA. 
   - Investigate LRT model selection of the random effects. Nested random effect models
   can be tested using LRT, and results from the "best" model is returned. The log-likelihood 
   is saved during each simulation, so the model selection can be done as a
   post-processing step; `summary.plcp_sim(..., model_selection = "FW", LRT_alpha = 0.25)`.
   
## Breaking canges
* `simulate(formula = x)` must now be created using the new functions `sim_formula`, or
`sim_formula_compare`, and can no longer be a named list or a character vector. 

## Bug fixes
* `summary.plcp_sim()` now show fixed effect `theta`s in the correct order, thanks to
GitHub user Johnzav888 (#10).

# Changes in version 0.3.0

## New features
* More flexible effect size specification. This version adds support for:
   - unstandardized effect sizes, e.g. `effect_size = 5`,
   - and Cohen's d effect sizes that are standardized 
   using either the pre- or posttest SD, or the random slope SD,
   e.g. `effect_size = cohend(0.5, "posttest_sd")`

## Other changes
* Support for lmerTest 3.0.

# Changes in version 0.2.0
## New features
* Analytical power calculations now support using Satterthwaite's degrees of
  freedom approximation.
* `Simulate.plcp` will now automatically create lme4 formulas if none is
  supplied, see `?create_lmer_formula`.
* You can now choose what alpha level to use.
* Treat cluster sizes as a random variable, `uneqal_clusters` now accepts
  a function indicating the distribution of cluster sizes, via the new argument
  `func`, e.g. `rpois` or `rnorm` could be used to draw cluster sizes.
* Expected power for designs with parameters that are random variables,
  can be calculated by averaging over multiple realizations, using the
  argument `R`.
* Support for parallel computations on Microsoft Windows, and in GUIs/interactive
  environments, using `parallel::makeCluster` (PSOCK). Forking is still used for
  non-interactive Unix environments.
  
## Improvements
* Calculations of the variance of the treatment effect is now much faster for
  designs with unequal clusters and/or missing data, when cluster sizes are
  large. The calculations now use the much faster implementation used by lme4.
* Cleaner print-methods for `plcp_multi`-objects.
* Multiple power calculations can now be performed in parallel, via the
  argument `cores`.
* `simulate.plcp_multi` now have more options for saving intermediate results.
* `print.plcp_multi_power` now has better support for subsetting via either [], 
head(), or subset().

## Breaking changes
* `icc_pre_subject` is now defined as `(u_0^2 + v_0^2) / (u_0^2 + v_0^2 + error^2)`,
instead of `(u_0^2) / (u_0^2 + v_0^2 + error^2)`. This would be the subject-level ICC, 
if there's no random slopes, i.e. correlation between time points for the same subject.
* `study_parameters()`: 0 and NA now means different things. If 0 is passed, the parameters
is kept in the model, if you want to remove it specify it as NA instead.
* `study_parameters()`: is now less flexible, but more robust. Previously a large
combination if raw and relative parameters could be combined, and the individual 
parameters was solved for. To make the function less bug prone and easier to maintain,
it is now only possible to specify the cluster-level variance components as relative values,
if the other parameters as passed as raw inputs. 

## Bug fixes and minor changes
* Output from `simulate_data()` now includes a column `y_c` that contains the full outcome vector,
without missing values added. This makes it easy to compare the complete and incomplete
data set, e.g. via `simulate()`.
* `simulate()` new argument `batch_progress` enables showing progress when doing
  multiple simulations.
* Bugfix for `summary.plcp_sim` where the wrong % convergence was calculated.
* Simulation function now accepts lme4 formulas containing "||".
* The cluster-level intercept variance is now also set to zero in the control
  group, when a partially nested design is requested.
* Fix incorrect error message from `study_parameters` when
  `icc_cluster_pre = NULL` and all inputs are standardized.
* Fix bug that would cause all slopes to be zero when `var_ratio` argument was
  passed a vector of values including a 0, e.g. `var_ratio = c(0, 0.1, 0.2)`.
* Bugfix for multi-sim objects that caused the wrong class the be used for,
  e.g. `res[[1]]$paras`, and thus the single simulation would not print
  correctly.
* Results from multi-sim objects can now be summarized for all random effects
  in the model.
* More support for summarizing random effects from partially nested formulas,
  e.g. `cluster_intercept` and `cluster_slope` is now correctly extracted from
  `(0 + treatment + treatment:time || cluster)`.
* When Satterthwaite's method fails the between clusters/subjects *dfs*
  are used to calculate *p*-values.
* `Power.plcp_multi` is now exported.
* `get_power.plcp_multi` now shows a progress bar.

R/VPC.r
# variance partitioning


#' Calculate the variance partitioning coefficient
#'
#' @param object An object created by \code{\link{study_parameters}}
#'
#' @details For partially nested studies, the VPC is calculated for the treatment group.
#'
#' @return a \code{data.frame} with class \code{plcp_VPC} containing the
#'  percentage of variance per level and time point. The column
#'  \code{between_clusters} is also the intraclass correlation for level three,
#'  i.e. the correlation between two subjects belonging to the same cluster at
#'  a specific time point. With random slopes in the model the variances per time point
#'  will be a quadratic function of time. \code{tot_var} is the
#'  percentage increase or decrease in total variance relative to baseline variance.
#'
#' The \code{plot} method returns a \code{ggplot2::ggplot} object.
#' @seealso \code{\link{plot.plcp_VPC}}
#'
#' @references Goldstein, H., Browne, W., & Rasbash, J. (2002).
#' Partitioning variation in multilevel models.
#' \emph{Understanding Statistics: Statistical Issues in Psychology, Education,
#' and the Social Sciences, 1}(4), 223-231.
#' @examples
#' paras <- study_parameters(n1 = 11,
#'                           n2 = 10,
#'                           n3 = 3,
#'                           T_end = 10,
#'                           icc_pre_subject = 0.5,
#'                           icc_pre_cluster = 0,
#'                           icc_slope = 0.05,
#'                           var_ratio = 0.03)
#'
#' res <- get_VPC(paras)
#' res
#'
#' # Plot
#' plot(res)
#' @export
get_VPC <- function(object) {
     UseMethod("get_VPC")
}

#' @rdname get_VPC
#' @export
get_VPC.plcp <- function(object) {
    paras <- NA_to_zero(object)
    u0 <- paras$sigma_subject_intercept
    u1 <- paras$sigma_subject_slope
    v0 <- paras$sigma_cluster_intercept
    v1 <- paras$sigma_cluster_slope
    v01 <- v0 * v1 * paras$cor_cluster
    error <- paras$sigma_error
    u01 <- paras$cor_subject * u0 * u1
    time <- get_time_vector(paras)
    tot_var <- (u0^2 + 2 * u01 * time + u1^2 * time^2 +
        v0^2 + 2 * v01 * time + v1^2 * time^2 + error^2)
    lvl3 <- (v0^2 + 2 * v01 * time + v1^2 * time^2) / tot_var
    lvl2 <- (u0^2 + 2 * u01 * time + u1^2 * time^2) / tot_var
    lvl1 <- error^2 / tot_var
    tot_var <- tot_var / tot_var[1]
    res <- data.frame(time,
        between_clusters = lvl3 * 100,
        between_subjects = lvl2 * 100,
        within_subjects = lvl1 * 100,
        tot_var = (tot_var - 1) * 100)
    class(res) <- append("plcp_VPC", class(res))

    res
}

#' @export
get_VPC.plcp_multi <- function(object) {
    warning("Multiple study designs used, only the first is shown")
    get_VPC(as.plcp(object[1, ]))
}


#' Plot method for \code{get_VPC}-objects
#'
#' @param x An object created with \code{\link{get_VPC}}
#' @param ... Optional arguments, currently ignored.
#'
#' @export

plot.plcp_VPC <- function(x, ...) {
     check_installed("ggplot2")
     res <- x
     res$tot_var <- NULL

     res <- stats::reshape(res, direction = "long",
                    varying = 2:4,
                    times = c("between_clusters",
                              "between_subjects",
                              "within_subjects"),
                    v.names = "proportion", timevar = "level")
     res$level <- factor(res$level, levels = c("between_clusters",
                                               "between_subjects",
                                               "within_subjects"),
                         labels = c("between-clusters (L3)",
                                    "between-subjects (L2)",
                                    "within-subjects (L1)"))

    p <- ggplot2::ggplot(res, ggplot2::aes_string("time", "proportion", color = "level", fill = "level")) +
        ggplot2::geom_line() +
        ggplot2::geom_point() +
        ggplot2::labs(title = "Variance partitioning",
               x = "Time point",
               y = "Percentage of total variance (%)")

    if(requireNamespace("ggsci", quietly = TRUE)) {
       p <- p + ggsci::scale_fill_d3() +
            ggsci::scale_color_d3()
    }
    p
}

#' Print method for \code{get_vpc}-objects
#' @param x Object created with \code{link{get_VPC}}
#' @param digits Number of digits to print
#' @param ... Optional arguments
#' @method print plcp_VPC
#' @export
print.plcp_VPC <- function(x, digits = 2, ...) {
    cat("# Percentage (%) of total variance at each level and time point\n")
    print.data.frame(x, digits = digits, scientific = FALSE, ...)
    cat("# 'tot_var' is the increase in variance compared to time = 0.\n")
    invisible(x)
}



# Standard deviations -----------------------------------------------------

#' Calculate the model implied standard deviations per time point
#'
#' @param object An object created by \code{\link{study_parameters}}
#' @param treatment \code{character}; either \code{"treatment"} or \code{"control"}.
#' Indicates for which group SDs should be calculated for. This only makes a difference
#' for 3-level partially nested designs.
#' @param n Optional; selects row n if \code{object} is a \code{data.frame} of
#' parameters
#'
#' @seealso \code{\link{get_VPC}}, \code{\link{get_correlation_matrix}}
#' @return \code{data.frame} with class \code{plcp_sds} containing the model
#' implied standard deviations per time point.
#'
#' @export
#'
#' @examples
#' paras <- study_parameters(n1 = 11,
#'                           n2 = 10,
#'                           n3 = 6,
#'                           T_end = 10,
#'                           icc_pre_subject = 0.5,
#'                           icc_pre_cluster = 0,
#'                           icc_slope = 0.05,
#'                           var_ratio = 0.03)
#'
#' get_sds(paras)
#'
#' # plot
#' plot(get_sds(paras))
#'
get_sds <- function(object, treatment = "treatment", n = 1) {
    if(!treatment %in% c("treatment", "control")) stop("Wrong 'treatment', allowed options are: 'treatment' or 'control'", call. = FALSE)
     UseMethod("get_sds")
}

#' @export
get_sds.plcp <- function(object, treatment = "treatment", n = NULL) {
    .p <- NA_to_zero(object)
    .p <- prepare_paras(.p)
    if(treatment == "treatment") {
        .p <- .p$treatment
    } else if(treatment == "control") {
        .p <- .p$control
    }
    .p$retention <- NULL
    .p$n2 <- NULL
    .p <- .p[c("sigma_subject_intercept",
                 "cor_subject",
                 "cor_cluster",
                 "sigma_subject_slope",
                 "sigma_cluster_slope",
                 "sigma_cluster_intercept",
                 "sigma_error",
                 "n1",
                 "T_end")]

    res <- do.call(get_sds_, .p)

     class(res) <- append(c("plcp_sds"), class(res))

     res
}

#' @export
get_sds.plcp_multi <- function(object, treatment = "treatment", n = 1) {
    get_sds.plcp(as.plcp(object[n, ]), treatment = treatment)

}

get_sds_ <- function(sigma_subject_intercept,
                     cor_subject,
                     sigma_subject_slope,
                     sigma_cluster_intercept,
                     sigma_cluster_slope,
                     cor_cluster,
                     sigma_error,
                     n1,
                     T_end) {
     time <-  seq(0, T_end, length.out = n1)
     u0 <- sigma_subject_intercept
     u1 <- sigma_subject_slope
     u01 <- u0 * u1 * cor_subject
     v0 <- sigma_cluster_intercept
     v1 <- sigma_cluster_slope
     v01 <- v0 * v1 * cor_cluster

     error <- sigma_error
     sds <- sqrt((u0^2 + 2*u01*time + u1^2*time^2 + v0^2 +
                      2 * v01*time + v1^2*time^2 + error^2))

     sds_lvl2 <- sqrt((u0^2 + 2*u01*time + u1^2*time^2 + v0^2 + error^2))

     res <- data.frame(time = time,
                       SD = sds,
                       SD_no_cluster_random_slope = sds_lvl2,
                       SD_no_random_slopes = sqrt(u0^2 + v0^2 + error^2))

     res
}


#' Plot method for \code{get_sds}-objects
#' @param x An object of class \code{plcp_sds}.
#' @param ... Optional arguments.
#' @export
plot.plcp_sds <- function(x, ...) {
     check_installed("ggplot2")
    .res <- x
     cs <- .res$SD_no_random_slopes[1]

     res <- .res
     res$time <- round(res$time,1)

     p <- ggplot2::ggplot(res, ggplot2::aes_string("time", "SD")) +
         ggplot2::geom_hline(ggplot2::aes_string(color = "'Random slopes = 0'",
                         yintercept = "SD_no_random_slopes")) +
         ggplot2::geom_line(ggplot2::aes(color = "With random slopes")) +
         ggplot2::geom_point(ggplot2::aes(color = "With random slopes")) +
         ggplot2::scale_x_continuous(breaks = unique(res$time)) +
         ggplot2::labs(y = "SD", x = "Time point",
               title = "SD per time point",
               color = "Model")
     # facet_grid(~cor_cluster + cor_subject, labeller = label_both)

     #  if(!is.nulls(facet)) p + facet_wrap(facet)

     p

}

#' Print method for \code{get_sds}-objects
#' @param x An object of class \code{plcp_sds}.
#' @param ... Optional arguments.
#' @export
#' @method print plcp_sds
print.plcp_sds <- function(x, digits = 2, ...) {
    print.data.frame(x, digits = digits, ...)
}


#
# CORRELATIONS LVL 2
#

#' Calculate the subject-level (ICC) correlations among time points
#'
#' @param object An object created by \code{\link{study_parameters}}
#'
#' @return A \code{n1} x \code{n1} \code{matrix} with the marginal subject-level
#' correlations between time points.
#' @details The correlation between time point \eqn{T_i} and \eqn{T_{i+1}} within
#' the same subject is also called the intraclass correlation (ICC) at level two.
#' If the random slopes are non-zero this ICC change over time.
#' @export
#'
#' @examples
#' paras <- study_parameters(n1 = 11,
#'                           n2 = 10,
#'                           n3 = 3,
#'                           T_end = 10,
#'                           icc_pre_subject = 0.5,
#'                           icc_pre_cluster = 0,
#'                           icc_slope = 0.05,
#'                           var_ratio = 0.03)
#' get_correlation_matrix(paras)
get_correlation_matrix <- function(object) {
    UseMethod("get_correlation_matrix")
}

#' @export
get_correlation_matrix.plcp <- function(object) {
    paras <- NA_to_zero(object)

    u0 <- paras$sigma_subject_intercept
    u1 <- paras$sigma_subject_slope
    v0 <- paras$sigma_cluster_intercept
    v1 <- paras$sigma_cluster_slope
    v01 <- v0 * v1 * paras$cor_cluster
    error <- paras$sigma_error
    u01 <- paras$cor_subject * u0 * u1
    time <- get_time_vector(paras)

    n1 <- paras$n1
    n2 <- paras$n2
    sx2 <- sum( (time - mean(time))^2)/n1

    X <- matrix(c(rep(1, n1), time), ncol = 2)
    Z <- X
    D <- matrix(c(u0^2, u01,
                  u01, u1^2), ncol = 2)
    D2 <- matrix(c(v0^2, v01,
                   v01, v1^2), ncol = 2)

    V <- Z %*% D %*% t(Z) + Z %*% D2 %*% t(Z) + error^2*diag(n1)


    V <- cov2cor(V)

    time_rounded <- round(time, 1)
    dimnames(V) <- list(time_rounded, time_rounded)

    class(V) <- append(class(V), "plcp_ICC2")

    V
}

#' @rdname get_correlation_matrix
#' @export
get_correlation_matrix.plcp_multi <- function(object) {
    warning("Multiple study designs used, only the first is shown")
    get_correlation_matrix.plcp(object[1, ])
}


#' Plot method for \code{get_correlation_matrix}-objects
#'
#' @param x An object created with \code{\link{get_correlation_matrix}}
#' @param ... Optional arguments, currently ignored.
#'
#' @export


plot.plcp_ICC2 <- function(x, ...) {
    check_installed("ggplot2")
    res <- as.data.frame(x)

    breaks <- 1:ncol(res)
    res <- reshape(res, varying = breaks,
                   v.names = "cor",
                   idvar = "time1",
                   timevar = "time2",
                   direction = "long")
    res$time1 <- res$time1
    res$time2 <- res$time2
    res$cor2 <- round(res$cor, 2)

    break_labs <- as.numeric(dimnames(x)[[1]])

    p <- ggplot2::ggplot(res, ggplot2::aes_string("time1", "time2", color = "cor", fill = "cor")) +
        ggplot2::geom_tile() +
        ggplot2::geom_text(ggplot2::aes_string(label = "cor2"), hjust = "center", color = "black") +
        ggplot2::scale_x_continuous(breaks = breaks, labels = break_labs) +
        ggplot2::scale_y_continuous(breaks = breaks, labels = break_labs) +
        ggplot2::labs(color = "Correlation", fill = "Correlation",
             x = "Time", y = "Time",
             title = "Subject-level correlation matrix") +
        ggplot2::theme_minimal()

    if(requireNamespace("viridis", quietly = TRUE)) {
        p <- p + viridis::scale_fill_viridis() +
            viridis::scale_color_viridis()
    }
    p

}

#' Print method for \code{get_correlation_matrix}-objects
#'
#' @param x An object created by \code{\link{get_correlation_matrix}}
#'
#' @param ... Optional arguments
#'
#' @method print plcp_ICC2
#' @export
print.plcp_ICC2 <- function(x, digits = 2, ...) {
    x <- unclass(x)
    print(round(x, digits = digits), ...)
}